### DISCUSSION TITLE
Discrepancy between inferred type and z.ZodType when default() is used

### DISCUSSION THREAD

**Discussion by danmichaelo** [colinhacks/zod#2972] [Q&A] [answered]
Given a type and a matching schema

```ts
type DurationRecord = {
  hours?: number | undefined;
  minutes: number;
};

const DurationRecordSchema: z.ZodType<DurationRecord> = z.object({
  hours: z.number().optional(),
  minutes: z.number().default(0),
}); 
```

I would expect no error, but instead I get the following typescript error:

```
Type 'ZodObject<{ hours: ZodOptional<ZodNumber>; minutes: ZodDefault<ZodNumber>; }, "strip", ZodTypeAny, { minutes: number; hours?: number | undefined; }, { ...; }>' is not assignable to type 'ZodType<DurationRecord, ZodTypeDef, DurationRecord>'.
  The types of '_input.minutes' are incompatible between these types.
    Type 'number | undefined' is not assignable to type 'number'.
      Type 'undefined' is not assignable to type 'number'.ts(2322)
```

If we instead infer the type from the schema, we get what *seems* to be the same type:

![image](https://github.com/colinhacks/zod/assets/434495/88950b4b-52d0-4014-be67-f8bc186e9fad)

**Comment by santosmarco** [comment, marked as answer]
This is not an issue, and the code is functioning as expected.

`z.ZodType<DurationRecord>` automatically sets both the output and input of the schema to `DurationRecord`. `ZodType` is defined as `ZodType<Output, Def, Input = Output>`, where `Input = Output` is specified at the end.

However, `ZodDefault` sets the input of `minutes` to possibly `undefined`. This is because when you set a default value, you are explicitly instructing Zod to accept `undefined` values and default them to `0`.

To fix this, you can try the following code:

```ts
type DurationRecord = {
  hours?: number;
  minutes: number;
};

type DurationRecordInput = {
  hours?: number;
  minutes?: number;
};

const DurationRecordSchema: z.ZodType<DurationRecord, z.ZodTypeDef, DurationRecordInput> = z.object({
  hours: z.number().optional(),
  minutes: z.number().default(0),
});
```
